morebits.js

```diff
@@ -1380,11 +1380,18 @@ Morebits.wiki.api.prototype = {
 
 		++Morebits.wiki.numberOfActionsLeft;
 
+		var queryString = $.param(this.query);
+		if (this.query.token) { // token should always be the last item in the query string (bug TW-B-0013)
+			queryString = queryString
+				.replace(/^(.*?)(&token=[^&]*)(.*)/, '$1$3$2') // shunt token from middle of query string to the end
+				.replace(/^(token=[^&]*)&(.*)/, '$2&$1'); // shunt token from beginning of query string to the end
+		}
+
```

Comment:
I wrote this initially as 
```js
	if (this.query.token) {
		var token = this.query.token;
		delete this.query.token;
		this.query.token = token;
	}
```
followed by $.param(this.query.token), but the `delete` operator is purported to cause performance impacts in V8, and isn't used anywhere else in Twinkle. Anyway, the two lines of ugly regex is much better than the 25 lines of ugly code that was there originally in `Morebits.queryString.create`!

Reply:
Much nicer!  The original is certainly over-complicated, but I think casting into an array is indeed the generic way to do it (sans jQuery and `morebits` of course).  Either way, this should work for our specific purposes, so w00t!

Could we make it even tighter, though, by relying on the fact that tokens end in `+\`?  Would something like the following be sufficient?

```javascript
queryString = $.param(this.query).replace(/^(.*?)(token=[^%]+%2B%5C)&(.*)/, '$1$3&$2')
```

It:
* matches until `%`
* only matches if *something* is present (not sure what happens if there's an empty token parameter isn't last, but that shouldn't happen)
* removes the trailing `&` and places one before the terminal `token`:
    * if `token` is first, the `?` remains for the new first parameter
    * if `token` is in the middle, `token`'s `&` remains for the upcoming parameter
    * if `token` is last, there won't be a terminal `&` to match and there's nothing to do anyway!

Reply:
Great work in condensing it down to a single regex replace!

> (token=[^%]+%2B%5C)

not so sure whether this is an improvement. While a nice idea, it seems better to me to leave it as 

> (token=[^&]*)

$.param(this.query) encodes the URI components so `&` can't occur as part of the token itself. So encountering an `&` means that the token part has ended. So we aren't really tightening up anything right?

Reply:
NOTE: The code here doesn't account for the other features of `Morebits.queryString.create` - omitting parameters with value `undefined` and converting array parameters to `|`-delimited strings. 

This needs to be resolved before merge.

Reply:
So that'd be `queryString = $.param(this.query).replace(/^(.*?)(token=[^&]+)&(.*)/, '$1$3&$2')`, right? (`+` rather than `*`)

Reply:
meh, using `*` means an empty `token=` also gets moved to the end, which is better right?